src/Agent.Listener/CommandSettings.cs

```diff
@@ -13,115 +13,40 @@ namespace Microsoft.VisualStudio.Services.Agent.Listener
-            Constants.Agent.CommandLine.Args.MachineGroupTags,
-            Constants.Agent.CommandLine.Args.Matrix,
-            Constants.Agent.CommandLine.Args.MonitorSocketAddress,
-            Constants.Agent.CommandLine.Args.NotificationPipeName,
-            Constants.Agent.CommandLine.Args.Password,
-            Constants.Agent.CommandLine.Args.Pool,
-            Constants.Agent.CommandLine.Args.ProjectName,
-            Constants.Agent.CommandLine.Args.ProxyPassword,
-            Constants.Agent.CommandLine.Args.ProxyUrl,
-            Constants.Agent.CommandLine.Args.ProxyUserName,
-            Constants.Agent.CommandLine.Args.SslCACert,
-            Constants.Agent.CommandLine.Args.SslClientCert,
-            Constants.Agent.CommandLine.Args.SslClientCertKey,
-            Constants.Agent.CommandLine.Args.SslClientCertArchive,
-            Constants.Agent.CommandLine.Args.SslClientCertPassword,
-            Constants.Agent.CommandLine.Args.StartupType,
-            Constants.Agent.CommandLine.Args.Token,
-            Constants.Agent.CommandLine.Args.Url,
-            Constants.Agent.CommandLine.Args.UserName,
-            Constants.Agent.CommandLine.Args.WindowsLogonAccount,
-            Constants.Agent.CommandLine.Args.WindowsLogonPassword,
-            Constants.Agent.CommandLine.Args.Work
-        };
-
-        // Commands.
-        public bool Configure => TestCommand(Constants.Agent.CommandLine.Commands.Configure);
-        public bool Remove => TestCommand(Constants.Agent.CommandLine.Commands.Remove);
-        public bool Run => TestCommand(Constants.Agent.CommandLine.Commands.Run);
-        public bool Warmup => TestCommand(Constants.Agent.CommandLine.Commands.Warmup);
-
-        // Flags.
-        public bool Commit => TestFlag(Constants.Agent.CommandLine.Flags.Commit);
-        public bool Diagnostics => TestFlag(Constants.Agent.CommandLine.Flags.Diagnostics);
-        public bool Help => TestFlag(Constants.Agent.CommandLine.Flags.Help);
-        public bool Unattended => TestFlag(Constants.Agent.CommandLine.Flags.Unattended);
-        public bool Version => TestFlag(Constants.Agent.CommandLine.Flags.Version);
-        public bool DeploymentGroup => TestFlag(Constants.Agent.CommandLine.Flags.MachineGroup) || TestFlag(Constants.Agent.CommandLine.Flags.DeploymentGroup);
-        public bool DeploymentPool => TestFlag(Constants.Agent.CommandLine.Flags.DeploymentPool);
-        public bool EnvironmentVMResource => TestFlag(Constants.Agent.CommandLine.Flags.Environment);
-        public bool GitUseSChannel => TestFlag(Constants.Agent.CommandLine.Flags.GitUseSChannel);
-        public bool RunOnce => TestFlag(Constants.Agent.CommandLine.Flags.Once);
+        private Dictionary<string, string> _envArgs = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+        private IPromptManager _promptManager;
+        private Tracing _trace;
+        public CommandArgs args;
```

Comment:
I think making args public is a bad idea. If we can simply hide the implementation detail in CommandSettings, then we can change the implementation anytime we want. Exposing the CommandArgs type here means that changing it in the future will be difficult. 
It looks like most of the properties are already exposed here, we just need to add the missing ones.

Reply:
Yes Exactly!

It's odd that there is a mix of functions to get values and then some accessing properties directly (pre-existing).  

Do you think it would be too hard to review in this PR if I wrapped everything in functions here?  That was my concern, but yeah I definitely agree with you.

Reply:
The best PR answer would be to do a PR first that adds the properties/functions here without changing the implementation. Then change the implementation in a separate PR. That would allow your reviewers to focus on the important stuff in each PR.

Reply:
I second this. I'd keep the CommandSettings constructor taking the string[] args and keep the command parsing in CommandSettings. Then CommandSettings has member variables for each of the commands and you don't need CommandArgs at all.

Reply:
@mjroghelia that is a great suggestion!  By moving the parsing logic into CommandSettings the changes to the unit tests are much more minimal and they can actually verify the parsing!